compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,340 @@ fn check_for_underconstrained_values_within_function(
+        BrilligTaintedIds {
+            arguments: HashSet::from_iter(arguments.iter().copied()),
+            results: results.iter().map(|result| HashSet::from([*result])).collect(),
+            root_results: HashSet::from_iter(results.iter().copied()),
+            results_constrained: HashSet::new(),
+        }
+    }
+
+    /// Add children of a given parent to the tainted value set
+    /// (for arguments one set is enough, for results we keep them
+    /// separate as the forthcoming check considers the call covered
+    /// if all the results were properly covered)
+    fn update_children(&mut self, parent: &ValueId, children: &[ValueId]) {
+        if self.arguments.contains(parent) {
+            self.arguments.extend(children);
+        }
+        for (i, result) in &mut self.results.iter_mut().enumerate() {
+            // Skip updating results already found covered
+            if self.results_constrained.contains(&i) {
+                continue;
+            }
+            if result.contains(parent) {
+                result.extend(children);
+            }
+        }
+    }
+
+    /// If Brillig call is properly constrained by the given ids, return true
+    fn check_constrained(&self) -> bool {
+        // If every result has now been constrained,
+        // consider the call properly constrained
+        self.results.len() == self.results_constrained.len()
+    }
+
+    /// Remember partial constraints (involving some of the results and an argument)
+    /// along the way to take them into final consideration
+    fn store_partial_constraints(&mut self, constrained_values: &HashSet<ValueId>) {
+        let mut results_involved: Vec<usize> = vec![];
+
+        // For a valid partial constraint, a value descending from
+        // one of the results should be constrained
+        for (i, result_descendants) in self.results.iter().enumerate() {
+            // Skip checking already covered results
+            if self.results_constrained.contains(&i) {
+                continue;
+            }
+            if result_descendants.intersection(constrained_values).next().is_some() {
+                results_involved.push(i);
+            }
+        }
+
+        // Also, one of the argument descendants should be constrained
+        // (skipped if there were no arguments, or if an actual result and not a
+        // descendant has been constrained, e.g. against a constant)
+        if !results_involved.is_empty()
+            && (self.arguments.is_empty()
+                || self.root_results.intersection(constrained_values).next().is_some()
+                || self.arguments.intersection(constrained_values).next().is_some())
+        {
+            // Remember the partial constraint
+            self.results_constrained.extend(&results_involved);
+            // We can clear the unneeded sets now
+            results_involved.iter().for_each(|i| self.results[*i].clear());
```

Comment:
This makes me think that instead of this:
```rust
    results: Vec<HashSet<ValueId>>,
    // Indices in `results` already constrained
    results_constrained: HashSet<usize>,
```
You could have something like
```rust
enum ResultStatus { 
  Unconstrained { descendants: HashSet<ValueId> },
  Constrained,
}
results: Vec<ResultStatus>
```
Just a thought, maybe there are other reasons to keep the two separate.

Reply:
Yeah thanks this would be indeed cleaner. Done